apps/condo/domains/acquiring/tasks/recurrent-payments/recurrent-payment-processing.js

```diff
@@ -0,0 +1,128 @@
+    } else {
+        return await setRecurrentPaymentAsFailed(context, recurrentPayment, errorMessage, errorCode)
+    }
+}
+
+async function process () {
+    logger.info({ msg: 'Start processing recurrent payment tasks' })
+
+    // prepare context
+    const { keystone } = await getSchemaCtx('RecurrentPaymentContext')
+    const context = await keystone.createContext({ skipAccessControl: true })
+
+    // prepare vars
+    const { pageSize } = paginationConfiguration
+    let offset = 0
+    let hasMorePages = true
+
+    // retrieve RecurrentPaymentContext page by page
+    while (hasMorePages) {
+        logger.info({ msg: `Processing recurrent payment page #${Math.floor(offset / pageSize)}` })
+        // get page (can be empty)
+        const page = await getReadyForProcessingPaymentsPage(context, pageSize, offset)
+
+        // process each page in parallel
+        await processArrayOf(page).inParallelWith(async (recurrentPayment) => {
+            try {
+                // first step create multi payment
+                const {
+                    registered,
+                    errorCode,
+                    errorMessage,
+                    multiPaymentId,
+                    directPaymentUrl,
+                    getCardTokensUrl,
+                } = await registerMultiPayment(context, recurrentPayment)
+
+                // there can be a case when all receipts already paid
+                if (registered) {
+                    // then create payment adapter to work with acquiring integration
+                    const adapter = new PaymentAdapter({
+                        multiPaymentId,
+                        directPaymentUrl,
+                        getCardTokensUrl,
+                    })
+
+                    // then proceed payment logic
+                    await processRecurrentPayment(context, recurrentPayment, adapter)
+                } else if (errorCode) {
+                    // error at registration
+                    // - disabled recurrent payment context
+                    // - exceed limit
+                    await setRecurrentPaymentAsFailed(context, recurrentPayment, errorMessage, errorCode)
+                } else {
+                    // nothing to pay case
+                    await setRecurrentPaymentAsSuccess(context, recurrentPayment)
+                }
+            } catch (error) {
+                const message = get(error, 'errors[0].message') || get(error, 'message') || JSON.stringify(error)
+                logger.error({ msg: 'Process recurrent payment error', message })
+                await setRecurrentPaymentAsFailed(
+                    context,
+                    recurrentPayment,
+                    message,
+                    PAYMENT_ERROR_UNKNOWN_CODE,
+                )
+            }
+        })
+
+        hasMorePages = page.length > 0
+        offset += pageSize
+    }
+    logger.info({ msg: 'End processing recurrent payment' })
```

Comment:
it's impossible to correlate the start and the end message. We have too match logs ...

Reply:
Finally, I'm going to keep use dictionary for log message: along with msg, pass jobId that can be used for logs correlation